### ISSUE TITLE
Eloquent\Collection::merge() cannot be used with Support\Collection, but type is not easy to determine

### ISSUE THREAD

**Issue by matt1357** [laravel/framework#19784]
- Laravel Version: 5.4.* -- introduced in 5.3

### Description:
Eloquent\Collection::merge($other) assumes getKey() exists on objects in $other. This seems a bit weird, but is really weird when combined with the fact that Eloquent\Collection::map() returns either an Eloquent\Collection or a Support\Collection.

As a toy example, I might write:
// Get the names of users matching some condition
$names1 = User::where('some_condition',true)->get()->map(function($u) { return $u->name; });
// Get the names of users matching some other condition
$names2 = User::where('other_condition',true)->get()->map(function($u) { return $u->name; });
// Get all the names.
$names = $names1->merge($names2);

This works, because after applying map, both collections are direct instances of Support\Collection... Unless $names1 is empty, then map preserves the type Eloquent\Collection.

So: when $names1 is empty, Eloquent\Collection::merge(Support\Collection $names2) is called, and an exception like "method: getKey() not found on type string" is thrown.

If $names1 is not empty, then this code works as one might expect.

To avoid falling in to this trap, I require intimate knowledge of both collection classes. I could do something like ->get()->toBase(), which does not seem adequate to me. 

This issue is similar to:
https://github.com/laravel/framework/issues/9163
But I think the strange behaviour of map() returning different types depending on whether the  Eloquent\Collection is empty or not make this a different issue.

TLDR:
User::where('some_condition',true)->get()->map(function($u) { return $u->name; })->merge(
  User::where('other_condition',true)->get()->map(function($u) { return $u->name; })
);
Throws: method: getKey() not found on type string, if no users have 'some_condition'
Succeeds: if any user has 'some_condition'

**Comment by themsaid** [comment]
You can use `toBase` to convert it to a normal collection before merging, since you're returning a string it shouldn't be an Eloquent collection anymore.

**Comment by matt1357** [comment]
>since you're returning a string it shouldn't be an Eloquent collection anymore.

But it is still an eloquent collection if the collection was empty. That's the problem.
